c/dao.c

```diff
@@ -0,0 +1,290 @@
+  unsigned char withdraw_dao[DAO_SIZE];
+
+  len = DAO_SIZE;
+  ret = extract_bytes(ns(Header_dao(deposit_header)), deposit_dao, &len);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len < 8) {
+    return ERROR_ENCODING;
+  }
+  len = DAO_SIZE;
+  ret = extract_bytes(ns(Header_dao(withdraw_header)), withdraw_dao, &len);
+  if (ret != CKB_SUCCESS) {
+    return ret;
+  }
+  if (len < 8) {
+    return ERROR_ENCODING;
+  }
+
+  uint64_t deposit_accumulate_rate = *((uint64_t *)(&deposit_dao[8]));
+  uint64_t withdraw_accumulate_rate = *((uint64_t *)(&withdraw_dao[8]));
+
+  volatile uint64_t occupied_capacity = 0;
+  len = 8;
+  ret = ckb_load_cell_by_field(((unsigned char *)&occupied_capacity), &len, 0,
+                               input_index, CKB_SOURCE_INPUT,
+                               CKB_CELL_FIELD_OCCUPIED_CAPACITY);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+
+  uint64_t counted_capacity = 0;
+  if (__builtin_usubl_overflow(original_capacity, occupied_capacity,
+                                &counted_capacity)) {
+    return ERROR_OVERFLOW;
+  }
+
+  __int128 withdraw_counted_capacity = ((__int128)counted_capacity) *
+                                       ((__int128)withdraw_accumulate_rate) /
+                                       ((__int128)deposit_accumulate_rate);
+
+  uint64_t withdraw_capacity = 0;
+  if (__builtin_uaddl_overflow(occupied_capacity,
+                                (uint64_t)withdraw_counted_capacity,
+                                &withdraw_capacity)) {
+    return ERROR_OVERFLOW;
+  }
+
+  *calculated_capacity = withdraw_capacity;
+  return CKB_SUCCESS;
+}
+
+int main(int argc, char *argv[]) {
+  int ret;
+  unsigned char script_hash[HASH_SIZE];
+  volatile uint64_t len = 0;
+
+  /*
+   * DAO has no arguments, this way we can ensure all DAO related scripts
+   * in a transaction is mapped to the same group.
+   */
+  if (argc != 1) {
+    return ERROR_WRONG_NUMBER_OF_ARGUMENTS;
+  }
+
+  len = HASH_SIZE;
+  ret = ckb_load_script_hash(script_hash, &len, 0);
+  if (ret != CKB_SUCCESS) {
+    return ERROR_SYSCALL;
+  }
+
+  size_t index = 0;
+  uint64_t input_capacities = 0;
+  while (1) {
+    int dao_input = 0;
+    volatile uint64_t capacity = 0;
+    len = 8;
+    ret = ckb_load_cell_by_field(((unsigned char *)&capacity), &len, 0, index,
+                                 CKB_SOURCE_INPUT, CKB_CELL_FIELD_CAPACITY);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    } else if (ret == CKB_SUCCESS) {
+      unsigned char current_script_hash[HASH_SIZE];
+      len = HASH_SIZE;
+      ret = ckb_load_cell_by_field(current_script_hash, &len, 0, index,
+                                   CKB_SOURCE_INPUT, CKB_CELL_FIELD_TYPE_HASH);
+      if ((ret == CKB_SUCCESS) &&
+          (memcmp(script_hash, current_script_hash, HASH_SIZE) == 0)) {
+        dao_input = 1;
+      }
+    } else if (ret == CKB_ITEM_MISSING) {
+      /* DAO Issuing input here, we can just skip it */
```

Comment:
ℹ️**Review Information**

Issuing input is a special `OutPoint` which is used as a placeholder to ensure that outputs total capacity <= inputs total capacity always hold. Script `dao.c` here verifies that the issued interest is correct.

https://github.com/nervosnetwork/ckb/blob/d470e5d804f3669d54ed53262a9a002d078a9984/core/src/transaction.rs#L162